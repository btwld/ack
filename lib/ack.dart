library;

export 'src/ack.dart';
