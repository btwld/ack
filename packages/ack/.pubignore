# Development and IDE files
melos_ack.iml
.claude/

# Build artifacts
build/
coverage/

# Local example (if any development examples exist)
example/
