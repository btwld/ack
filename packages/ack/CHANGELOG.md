## 0.2.0-beta.1 (2025-05-03)

* See [release notes](https://github.com/btwld/ack/releases/tag/v0.2.0-beta.1) for details.


## 0.2.0-beta.1

 - Bump "ack" to `0.2.0-beta.1`.

## 0.2.0

 - Bump "ack" to 0.2.0 with improved SchemaModel API and enhanced string validation

## 0.2.0

### Breaking Changes

- **SchemaModel API**: Completely redesigned the SchemaModel API for a more intuitive and streamlined experience
  - Changed constructor parameter from `Map<String, dynamic>` to `Object?` for better flexibility
  - Added automatic validation during construction
  - Added `isValid` getter and `getErrors()` method for easier error handling
  - Added abstract `getSchema()` method to define the schema
  - Removed `parse()`, `fromValidated()`, `validateMap()`, `validate()`, `createFromMap()`, and `initialize()` methods

### Improvements

- Improved error handling with more detailed error messages
- Better type safety with `Object?` instead of `dynamic`
- Simplified API with fewer methods and more intuitive usage
- **String Validation**: Enhanced string validation with improved methods
  - Changed `matches()` to validate full string patterns (anchored with ^ and $)
  - Added `contains()` method for partial string matching

## 0.1.2

- Improved error messages

## 0.1.1

-  Added deprecations

## 0.1.0

-  Error and exception improvements
-  Reworked validation and constraint workflows
-  Improved testing

## 0.0.2

-  Better JSON response parsing for OpenApiSchemaConverter

## 0.0.1

-  Initial version.