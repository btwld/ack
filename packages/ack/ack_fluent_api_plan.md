# Ack Fluent API Refactor Plan

This document outlines the plan to implement a fluent, Zod-inspired API for the Ack validation library. The work is broken down by schema type and is based on a detailed analysis of our current architecture and the capabilities of the Zod library.

---

### **Guiding Principles**

1.  **Universal Methods go on the Base Class**: Any capability that can apply to *any* schema (like making it nullable or adding a description) should be a method on the base `AckSchema`.
2.  **Specific Methods go in Extensions**: Capabilities that only make sense for a specific type (e.g., email validation for strings) should be implemented as extension methods on that specific schema class (`StringSchema`, `ListSchema`, etc.). This keeps the core classes clean.
3.  **Unsupported Features**: We will clearly mark features from Zod that we cannot or should not implement right now due to significant architectural differences (e.g., transformations, async validation).

---

### **1. `AckSchema` (Base Class Methods)**

These methods will be added directly to the `AckSchema<T>` class because they are universal.

| Method Signature | Zod Equivalent | Why Here? | Implementation Notes |
| :--- | :--- | :--- | :--- |
| `nullable()` | `.nullable()` | Makes any schema type nullable. This is the most fundamental and universal operation. | Already implemented via the `.nullable()` extension. No new work needed. |
| `default(T value)` | `.default(value)` | Sets a default value if the input is `null` or `undefined`. A universal concept. | Already implemented via the `defaultValue` constructor parameter. The `copyWith` pattern allows us to set it fluently. |
| `description(String text)` | `.describe(text)` | Adds a description for documentation/error message purposes. Universal. | Already implemented via the `description` constructor parameter. |

**Conclusion:** The most fundamental universal methods are already supported by our new architecture.

---

### **2. `StringSchema` Extensions**

These will be implemented in a new file: `lib/src/schemas/extensions/string_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | String-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | String-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | String-specific length validation. | **Can Support Now** |
| `email()` | `.email()` | String-specific format validation. | **Can Support Now** |
| `url()` | `.url()` | String-specific format validation. | **Can Support Now** |
| `uuid()` | `.uuid()` | String-specific format validation. | **Can Support Now** |
| `cuid()` | `.cuid()` | String-specific format validation. | **Can Support Now** (less common, maybe lower priority) |
| `datetime()` | `.datetime()` | String-specific format validation. | **Can Support Now** |
| `regex(RegExp re)` | `.regex(re)` | String-specific pattern matching. | **Can Support Now** |
| `startsWith(String s)` | `.startsWith(s)` | String-specific content validation. | **Can Support Now** |
| `endsWith(String s)` | `.endsWith(s)` | String-specific content validation. | **Can Support Now** |
| `trim()` | `.trim()` | **Transformation**. | **Cannot Support Now**. Our architecture validates, it doesn't transform the input value. |
| `toLowerCase()` | `.toLowerCase()` | **Transformation**. | **Cannot Support Now**. |
| `toUpperCase()` | `.toUpperCase()` | **Transformation**. | **Cannot Support Now**. |

---

### **3. `IntegerSchema` & `DoubleSchema` Extensions**

These will be in `lib/src/schemas/extensions/num_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `gt(num n)` / `greaterThan(num n)` | `.gt(n)` | Numeric comparison. | **Can Support Now** |
| `gte(num n)` / `min(num n)` | `.gte(n)` | Numeric comparison. | **Can Support Now** |
| `lt(num n)` / `lessThan(num n)` | `.lt(n)` | Numeric comparison. | **Can Support Now** |
| `lte(num n)` / `max(num n)` | `.lte(n)` | Numeric comparison. | **Can Support Now** |
| `isPositive()` | `.positive()` | Numeric sign check. | **Can Support Now** |
| `isNegative()` | `.negative()` | Numeric sign check. | **Can Support Now** |
| `isNonPositive()` | `.nonpositive()` | Numeric sign check. | **Can Support Now** |
| `isNonNegative()` | `.nonnegative()` | Numeric sign check. | **Can Support Now** |
| `multipleOf(num n)` / `step(num n)` | `.multipleOf(n)` | Numeric divisibility check. | **Can Support Now** |
| `finite()` | `.finite()` | Checks if a number is finite. | **Can Support Now** (More relevant for `DoubleSchema`) |
| `safe()` | `.safe()` | Checks if an integer is a "safe" integer. | **Can Support Now** (For `IntegerSchema`) |

---

### **4. `ListSchema` Extensions**

These will be in `lib/src/schemas/extensions/list_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | List-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | List-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | List-specific length validation. | **Can Support Now** |
| `nonempty()` | `.nonempty()` | A convenient shorthand for `.minLength(1)`. | **Can Support Now** |

---

### **5. `ObjectSchema` Extensions**

This is where things get more complex. Our `ObjectSchema` is powerful but doesn't have all of Zod's dynamic capabilities.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `pick(List<String> keys)` | `.pick()` | **Structural Transformation**. | **Cannot Support Now**. This creates a *new* schema type by altering the properties. Our `copyWith` can't do this. |
| `omit(List<String> keys)` | `.omit()` | **Structural Transformation**. | **Cannot Support Now**. Same reason as `pick`. |
| `partial()` | `.partial()` | **Structural Transformation**. | **Cannot Support Now**. This changes the `required` status of all properties, which would require deep schema cloning. |
| `deepPartial()` | `.deepPartial()` | **Structural Transformation**. | **Cannot Support Now**. Even more complex than `partial`. |
| `merge(ObjectSchema other)` | `.merge()` | Combines two object schemas. | **Can Support Now (with limitations)**. We can implement a version of this that merges properties and `required` fields. |
| `passthrough()` | `.`passthrough()` | Allows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: true` flag. We could add a fluent method for it. |
| `strict()` | `.strict()` | Disallows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: false` flag. We could add a fluent method for it. |
| `catchall(AckSchema valueSchema)` | `.catchall()` | Validates all unknown keys against a schema. | **Cannot Support Now**. Our current implementation only checks for the *presence* of additional properties, not their *values*. |

---

### **6. Features We Cannot Support (The "Why")**

*   **Transformations (`.trim()`, `.toLowerCase()`, `.transform()`, `.refine()`)**:
    *   **Reason:** Our architecture has a clear separation of concerns: a schema's job is to **validate** an input and return a `SchemaResult` with either the original (but correctly typed) value or an error. It does not **mutate or transform** the input value. Zod's `.transform()` and `.refine()` fundamentally change the output value, which is a different paradigm. Supporting this would require a major redesign of the `parseAndValidate` flow.

*   **Structural Transformations (`.pick()`, `.omit()`, `.partial()`)**:
    *   **Reason:** These methods in Zod dynamically create *new* schema definitions at runtime by introspecting and cloning existing schemas. Our `copyWith` pattern is great for creating a modified *copy*, but it's not designed for this level of dynamic, structural alteration (e.g., removing keys from the `properties` map and the `requiredProperties` list simultaneously). It's possible, but would require significantly more complex logic in the `ObjectSchema`'s `copyWith` method.

*   **Asynchronous Validation (`.refineAsync()`)**:
    *   **Reason:** Our entire validation pipeline (`validate` -> `parseAndValidate` -> `SchemaResult`) is synchronous. Introducing `async` validation would require making the entire pipeline return `Future<SchemaResult>`, which is a massive and breaking change across the entire library. 