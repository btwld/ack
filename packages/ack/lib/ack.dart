/// ACK - A Comprehensive Kotlin-inspired validation library for Dart
library;

// Constraints
export 'src/constraints/constraint.dart';
// Context
export 'src/context.dart';
export 'src/schemas/extensions/list_schema_extensions.dart';
export 'src/schemas/extensions/numeric_extensions.dart';
export 'src/schemas/extensions/object_schema_extensions.dart';
// Extensions
export 'src/schemas/extensions/string_schema_extensions.dart';
// Core schemas
export 'src/schemas/schema.dart';
export 'src/validation/ack_exception.dart';
export 'src/validation/schema_error.dart';
// Validation results
export 'src/validation/schema_result.dart';
