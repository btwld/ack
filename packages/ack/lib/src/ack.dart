import 'schemas/extensions/string_schema_extensions.dart';
import 'schemas/schema.dart';

/// The main entry point for creating schemas with the Ack validation library.
///
/// Provides a fluent API for creating various schema types.
final class Ack {
  /// Creates a string schema.
  static StringSchema string() => const StringSchema();

  /// Creates a literal string schema that only accepts the exact [value].
  /// Similar to Zod's `z.literal("value")`.
  static StringSchema literal(String value) => string().literal(value);

  /// Creates an integer schema.
  static IntegerSchema integer() => const IntegerSchema();

  /// Creates a double schema.
  static DoubleSchema double() => const DoubleSchema();

  /// Creates a boolean schema.
  static BooleanSchema boolean() => const BooleanSchema();

  /// Creates an object schema with the given properties.
  static ObjectSchema object(
    Map<String, AckSchema> properties, {
    List<String> required = const [],
    bool additionalProperties = false,
  }) =>
      ObjectSchema(
        properties.cast(),
        required: required,
        additionalProperties: additionalProperties,
      );

  /// Creates a discriminated object schema for polymorphic validation.
  static DiscriminatedObjectSchema discriminated({
    required String discriminatorKey,
    required Map<String, ObjectSchema> schemas,
  }) =>
      DiscriminatedObjectSchema(
        discriminatorKey: discriminatorKey,
        schemas: schemas,
      );

  /// Creates a list schema with the given item schema.
  static ListSchema<T> list<T extends Object>(AckSchema<T> itemSchema) =>
      ListSchema(itemSchema);

  /// Creates an enum schema for validating enum values.
  static EnumSchema<T> enumValues<T extends Enum>(List<T> values) =>
      EnumSchema(values: values);

  /// Creates a schema that can be one of many types.
  static AnyOfSchema anyOf(List<AckSchema> schemas) => AnyOfSchema(schemas);
}
