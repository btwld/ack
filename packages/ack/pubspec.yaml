name: ack
description: A simple validation library for Dart
version: 0.2.0
repository: https://github.com/btwld/ack

environment:
  sdk: ^3.3.0

dependencies:
  meta: ^1.15.0

dev_dependencies:
  coverage: ^1.11.1
  dart_code_metrics_presets: ^2.22.0
  lints: ^5.0.0
  test: ^1.24.0
