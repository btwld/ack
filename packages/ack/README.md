# **ACK**
![GitHub stars](https://img.shields.io/github/stars/btwld/ack?style=for-the-badge&logo=GitHub&logoColor=black&labelColor=white&color=dddddd)
[![Pub Version](https://img.shields.io/pub/v/ack?label=version&style=for-the-badge&logo=dart&logoColor=3DB0F3&labelColor=white&color=3DB0F3)](https://pub.dev/packages/ack/changelog)
[![Pub Points](https://img.shields.io/pub/points/ack?style=for-the-badge&logo=dart&logoColor=3DB0F3&label=Points&labelColor=white&color=3DB0F3)](https://pub.dev/packages/ack/score)
[![All Contributors](https://img.shields.io/github/contributors/btwld/ack?style=for-the-badge&color=018D5B&labelColor=004F32)](https://github.com/btwld/ack/graphs/contributors)
[![MIT License](https://img.shields.io/github/license/btwld/ack?style=for-the-badge&color=FF2E00&labelColor=CB2500)](https://opensource.org/licenses/mit-license.php)
![Codecov](https://img.shields.io/codecov/c/github/btwld/ack?style=for-the-badge&color=FFD43A&labelColor=F3BE00)

ACK provides a fluent, unified schema-building solution for Dart and Flutter applications. It delivers clear constraints, descriptive error feedback, and powerful utilities for validating forms, AI-driven outputs, and JSON or CLI arguments.

See the full documentation at [docs.page/btwld/ack](https://docs.page/btwld/ack).

## Use Cases and Key Benefits

- Validates diverse data types with customizable constraints
- Converts into OpenAPI Specs for LLM function calling and structured response support
- Offers a fluent API for intuitive schema building
- Provides detailed error reporting for validation failures

## Installation

Add ACK to your `pubspec.yaml`:

```bash
dart pub add ack
```

## Usage Overview

ACK provides schema types to validate different kinds of data. You can customize each schema with constraints, nullability, strict parsing, default values, and more using a fluent API.

### String Schema

Validates string data, with constraints like minimum length, maximum length, non-empty checks, regex patterns, and more.

**Example**:

```dart
import 'package:ack/ack.dart';

final schema = Ack.string
  .minLength(5)
  .maxLength(10)
  .notEmpty()
  .nullable(); // Accepts null

final result = schema.validate('hello');
if (result.isOk) {
  print(result.getOrNull()); // "hello"
}
```

### Integer Schema

Validates integer data. Constraints include min/max values, exclusive bounds, and multiples.

**Example**:

```dart
final schema = Ack.int
    .min(0)
    .max(100)
    .multipleOf(5);

final result = schema.validate(25);
```

### Double Schema

Similar to Integer Schema, but for doubles:

```dart
final schema = Ack.double
    .min(0.0)
    .max(100.0)
    .multipleOf(0.5);

final result = schema.validate(25.5);
```

### Boolean Schema

Validates boolean data:

**Example**:

```dart
final schema = Ack.boolean.nullable();

final result = schema.validate(true);
```

This schema accepts boolean values or null.

### List Schema

Validates lists of items, each item validated by an inner schema:

**Example**:

```dart
final itemSchema = Ack.string.minLength(3);
final listSchema = Ack.list(itemSchema)
    .minItems(2)
    .uniqueItems();

final result = listSchema.validate(['abc', 'def']);
```

### Object Schema

Validates `Map<String, Object?>` with property definitions and constraints on required fields, additional properties, etc.

**Example**:

```dart
final schema = Ack.object(
  {
    'name': Ack.string.minLength(3),
    'age': Ack.int.min(0).nullable(),
  },
  required: ['name'],
);

final result = schema.validate({'name': 'John'});
```

This schema requires a "name" property (string, min length 3) and allows an optional "age" property (integer >= 0), with at least one property.

### Discriminated Union Schema

Validates objects where the structure depends on a specific "discriminator" field.

**Example**:

```dart
final schema = Ack.discriminated(
  discriminatorKey: 'type',
  schemas: {
    'user': Ack.object({
      'type': Ack.string,
      'name': Ack.string,
    }),
    'admin': Ack.object({
      'type': Ack.string,
      'level': Ack.int,
    }),
  },
);

final userResult = schema.validate({'type': 'user', 'name': 'Alice'}); // OK
final adminResult = schema.validate({'type': 'admin', 'level': 5});   // OK
final invalidResult = schema.validate({'type': 'guest'});            // Fail
```

## Additional Features

### Strict Parsing

For scalar schemas (String, Integer, Double, Boolean), ACK can parse strings or numbers into the correct type if strict is false (the default). If you set strict, the schema only accepts an already-correct type.

```dart
// By default, Ack.int will accept "123" and parse it to 123.
final looseSchema = Ack.int;
print(looseSchema.validate("123").isOk); // true

// If you require strictly typed ints (no string parsing):
final strictSchema = Ack.int.strict();
print(strictSchema.validate("123").isOk); // false
print(strictSchema.validate(123).isOk);   // true
```

### Default Values

You can set a default value so that if validation fails or if the user provides null, the schema returns the default:

```dart
// Setting default value in the constructor:
final schema = Ack.string(
  defaultValue: 'Guest',
  nullable: true,
).minLength(3);

// This fails the minLength check, but returns the default "Guest"
final result = schema.validate('hi');
print(result.getOrNull()); // "Guest"

final nullResult = schema.validate(null);
print(nullResult.getOrNull()); // "Guest"
```

> **Important**: If the parsed value is invalid or null, but a default value is present, ACK will return Ok(defaultValue) instead of failing.

### Custom Constraints

You can extend `ConstraintValidator<T>` or `OpenApiConstraintValidator<T>` to create your own validation rules. For example:

```dart
class OnlyFooStringValidator extends OpenApiConstraintValidator<String> {
  const OnlyFooStringValidator();

  @override
  String get name => 'only_foo';
  @override
  String get description => 'String must be "foo" only';

  @override
  bool isValid(String value) => value == 'foo';

  @override
  ConstraintError onError(String value) {
    return buildError(
      message: 'Value "$value" is not "foo".',
      context: {'value': value},
    );
  }

  // If you want this constraint to appear in OpenAPI:
  @override
  Map<String, Object?> toSchema() {
    // Typically you'd put `"enum": ["foo"]`, or similar
    return {
      'enum': ['foo'],
      'description': 'Must be exactly "foo"',
    };
  }
}

// Using it:
final schema = Ack.string.withConstraints([OnlyFooStringValidator()]);
final result = schema.validate("bar"); // Fails validation
```

### Custom OpenAPI Constraints

When you implement `OpenApiConstraintValidator<T>`, your custom validator's `toSchema()` output is automatically merged into the final JSON schema. This means you can add fields like `pattern`, `enum`, `minimum`, etc., as recognized by OpenAPI or JSON Schema.

```dart
// Example usage with the built-in OpenApiSchemaConverter
final converter = OpenApiSchemaConverter(schema: schema);
print(converter.toJson());
```

The library merges all constraints' `toSchema()` results, so you get a single cohesive OpenAPI spec for your entire schema.

### Fluent API

ACK's fluent API lets you chain methods:

```dart
final schema = Ack.int
  .min(0)
  .max(100)
  .multipleOf(5)
  .nullable() // accept null
  .strict();  // require actual int type
```

### Validation and Parsing Methods

Ack provides multiple methods for validating and parsing data, each offering different error handling patterns to suit your coding style:

#### validate() - Result Pattern

The `validate()` method uses a Result pattern, returning a `SchemaResult<T>` which can be either:
- `Ok<T>`: Contains the validated value, accessible via `getOrNull()` or `getOrThrow()`
- `Fail<T>`: Contains validation errors, accessible via `getErrors()`

This approach gives you fine-grained control over error handling and access to the full error details.

```dart
// Schema for a user object with validation rules
final userSchema = Ack.object({
  'name': Ack.string.minLength(2),
  'email': Ack.string.email()
}, required: ['name', 'email']);

// Validate user data
final result = userSchema.validate({
  'name': 'John',
  'email': 'not-an-email'
});

if (result.isOk) {
  // Access the validated data
  final validData = result.getOrThrow();
  print("Valid user: ${validData['name']}");
} else {
  // Access detailed validation errors
  final errors = result.getErrors();
  print("Validation failed: ${errors.name}");
  
  // You can inspect specific errors
  for (final issue in errors.issues) {
    print(" - ${issue.path}: ${issue.message}");
  }
}
```

#### parse() - Exception Pattern

The `parse()` method provides a more direct approach using exceptions for error handling. It returns the validated data directly if validation succeeds, or throws an `AckException` if validation fails.

```dart
try {
  // Return the validated data directly
  final validatedData = userSchema.parse({
    'name': 'John Doe',
    'email': 'john@example.com'
  });
  
  // Use the validated data immediately
  print("Valid user: ${validatedData['name']}");
} catch (e) {
  // Handle validation failure with exception
  print("Validation failed: $e");
}
```

#### tryParse() - Null Safety Pattern

The `tryParse()` method uses Dart's null safety for error handling. It returns the validated data if validation succeeds, or `null` if validation fails.

```dart
// Returns validated data or null on validation failure
final maybeData = userSchema.tryParse({
  'name': 'John Doe',
  'email': 'invalid-email' // This will fail validation
});

if (maybeData != null) {
  // Use the validated data
  print("Valid user: ${maybeData['name']}");
} else {
  // Handle validation failure
  print("Validation failed");
}
```

#### validateOrThrow() - Exception with Return Value

The `validateOrThrow()` method throws an exception on error, but also returns the validated data for immediate use:

```dart
try {
  // Validate and get data in one step
  final validData = userSchema.validateOrThrow({
    'name': 'John Doe',
    'email': 'john@example.com'
  });
  
  // Use the validated data immediately
  print("Valid user: ${validData['name']}");
} catch (e) {
  // Handle validation failure with exception
  print("Validation failed: $e");
}
```

#### Choosing the Right Approach

- Use `validate()` when you need detailed error information or want to handle valid and invalid cases in the same code path
- Use `parse()` when you prefer exception-based error handling and want to get the validated data directly
- Use `tryParse()` when you prefer null-safety-based error handling for cleaner code without try/catch blocks
- Use `validateOrThrow()` when you want to use exception handling but also need direct access to the validated data

### OpenAPI Spec

ACK can generate OpenAPI schema definitions from your schemas, aiding in API documentation or code generation.

```dart
final schema = Ack.object({
  'name': Ack.string
    .minLength(2)
    .maxLength(50),
  'age': Ack.int
    .min(0)
    .max(120),
}, required: ['name', 'age']);

final converter = OpenApiSchemaConverter(schema: schema);
final openApiSchema = converter.toSchema();

print(openApiSchema);

/* Returns schema like:
{
  "type": "object",
  "required": ["name", "age"],
  "properties": {
    "name": {
      "type": "string",
      "minLength": 2,
      "maxLength": 50
    },
    "age": {
      "type": "integer",
      "minimum": 0,
      "maximum": 120
    }
  }
}
*/
```

#### Working with Limited LLM OpenAPI Support

> [!TIP]
> When an LLM has limited support for OpenAPI function calling schemas but *can* guarantee valid JSON output, you can embed the schema definition directly within the prompt using `toResponsePrompt()`. This allows you to still validate the LLM's JSON output using `parseResponse()` against your ACK schema.

```dart
final schema = Ack.object(
  {
    'name': Ack.string.minLength(2).maxLength(50),
    'age': Ack.int.min(0).max(120),
  },
  required: ['name', 'age'],
);

final converter = OpenApiSchemaConverter(schema: schema);

// Build a prompt for the LLM that includes the schema
final prompt = '''
You are a helpful assistant. Please provide information about a person following this schema:

${converter.toResponsePrompt()}
''';

/* Will output:
<schema>
{
  "type": "object",
  "required": ["name", "age"],
  "properties": {
    "name": {
      "type": "string",
      "minLength": 2,
      "maxLength": 50
    },
    "age": {
      "type": "integer",
      "minimum": 0,
      "maximum": 120
    }
  },
  "additionalProperties": false
}
</schema>

Your response should be valid JSON, that follows the <schema> and formatted as follows:

<response>
{valid_json_response}
</response>
<stop_response>
*/

// Simulated LLM response
final llmResponse = '''
Here is the person's information:
<response>
{
  "name": "John Smith",
  "age": 35
}
</response>
''';

final jsonPayload = converter.parseResponse(llmResponse);

print(jsonPayload);
```


### Error Handling with SchemaResult

Every call to `.validate(value)` returns a `SchemaResult<T>` object, which is either `Ok<T>` or `Fail<T>`:
- `Ok`: Access the data via `getOrNull()` or `getOrThrow()`
- `Fail`: Inspect `getErrors()` for a list of `SchemaError` describing the failures

### Quick Reference

1. Validation & Parsing Methods:
   - `validate(value)` → `SchemaResult<T>` (Result pattern)
   - `parse(value)` → `T` or throws `AckException` (Exception pattern)
   - `tryParse(value)` → `T?` returns null on validation failure (Null safety pattern)
   - `validateOrThrow(value)` → `T` or throws `AckException` on errors
2. Fluent Methods:
   - `nullable()`
   - `strict()`
   - `withConstraints([ ... ])` 
3. Default Values: Provide `defaultValue: T?` directly in the schema constructor or via `.call(defaultValue: X)`.
4. Custom Constraints: Extend `ConstraintValidator<T>` or `OpenApiConstraintValidator<T>` to add your own logic.
5. OpenAPI: Use `OpenApiSchemaConverter(schema: yourSchema).toSchema()` (or `.toJson()`) to generate specs.

Happy validating with ACK!

## SchemaModel API

ACK provides a `SchemaModel` base class for creating schema-based models with automatic validation.

```dart
// Define a schema model manually (or use code generation)
class UserSchema extends SchemaModel {
  UserSchema(Object? data) : super(data);

  @override
  AckSchema getSchema() {
    return Ack.object({
      'name': Ack.string.minLength(2),
      'email': Ack.string.email(),
      'age': Ack.int.min(0).nullable(),
    }, required: ['name', 'email']);
  }
  
  // Getters for typed access to properties
  String get name => getValue<String>('name')!;
  String get email => getValue<String>('email')!;
  int? get age => getValue<int?>('age');
}

// Using the schema model - Constructor approach
final userData = {
  'name': 'John Doe',
  'email': 'john@example.com',
  'age': 30,
};

final userSchema = UserSchema(userData);

// Check if the schema is valid
if (userSchema.isValid) {
  // Access properties directly from the schema
  print('User: ${userSchema.name}, ${userSchema.email}, ${userSchema.age}');
  
  // Create a model instance manually
  final user = User(
    name: userSchema.name,
    email: userSchema.email,
    age: userSchema.age,
  );
} else {
  // Handle validation errors
  print('Validation errors: ${userSchema.getErrors()}');
}

// Creating and validating schemas
try {
  // Create schema and validate
  final userSchema = UserSchema(userData);
  if (userSchema.isValid) {
    print('User: ${userSchema.name}, ${userSchema.email}, ${userSchema.age}');
  } else {
    print('Validation failed: ${userSchema.getErrors()}');
  }
} catch (e) {
  print('Unexpected error: $e');
}

// Create your model class however you prefer
class User {
  final String name;
  final String email;
  final int? age;
  
  User({required this.name, required this.email, this.age});
}
```


## Documentation

For detailed guides on using Ack effectively, check out the documentation:

- [SchemaModel API](https://docs.page/leofarias/ack/core-concepts/schema-model-class) - Learn how to use the SchemaModel API

## License