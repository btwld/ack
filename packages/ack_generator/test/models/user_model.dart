import 'package:ack/ack.dart';

/// Annotate your model class with @AckModel
@Schema(
  description: 'A user model with validation',
  additionalProperties: true,
  additionalPropertiesField: 'metadata',
)
class User {
  /// Email must be valid
  @IsEmail()
  final String email;

  /// Name cannot be empty
  @IsNotEmpty()
  final String name;

  /// Age is optional
  final int? age;

  /// Password must be at least 8 characters and is required
  @Required()
  @MinLength(8)
  final String? password;

  /// Address must be provided
  @Required()
  final Address address;

  /// Additional properties are stored here
  final Map<String, dynamic> metadata;

  /// Constructor parameters determine which fields are required
  User({
    required this.email,
    required this.name,
    this.age,
    this.password,
    required this.address,
    Map<String, dynamic>? metadata,
  }) : metadata = metadata ?? {};

  /// String representation of user
  @override
  String toString() {
    return 'User{email: $email, name: $name, age: $age, password: ${password != null ? "****" : "null"}, address: $address, metadata: $metadata}';
  }
}

/// Address model with its own validation
@Schema()
class Address {
  /// Street cannot be empty
  @IsNotEmpty()
  final String street;

  /// City cannot be empty
  @IsNotEmpty()
  final String city;

  /// Zip is optional
  @Nullable()
  final String? zip;

  /// Constructor determines required fields
  Address({
    required this.street,
    required this.city,
    this.zip,
  });

  /// String representation of address
  @override
  String toString() {
    return 'Address{street: $street, city: $city, zip: $zip}';
  }
}
