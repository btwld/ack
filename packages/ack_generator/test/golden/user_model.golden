// GENERATED CODE - DO NOT MODIFY BY HAND

// ignore_for_file: unnecessary_this, unnecessary_new, unnecessary_const, prefer_collection_literals
// ignore_for_file: lines_longer_than_80_chars, unnecessary_null_checks, non_constant_identifier_names

part of 'user_model.dart';

// **************************************************************************
// AckSchemaGenerator
// **************************************************************************

/// Generated schema for User
/// A user model with validation
class UserSchema extends SchemaModel<User> {
  UserSchema([Object? value = null]) : super(value);

  static final ObjectSchema schema = Ack.object(
    {
      'email': Ack.string.email(),
      'name': Ack.string.notEmpty(),
      'age': Ack.int.nullable(),
      'password': Ack.string.minLength(8).nullable(),
      'address': AddressSchema.schema,
    },
    required: ['email', 'name', 'password', 'address'],
    additionalProperties: true,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<User, UserSchema>(
      (data) => UserSchema(data),
    );
    // Register schema dependencies
    AddressSchema.ensureInitialize();
  }

  @override
  AckSchema getSchema() => schema;

  String get email => getValue<String>('email')!;

  String get name => getValue<String>('name')!;

  int? get age => getValue<int>('age');

  String? get password => getValue<String>('password');

  AddressSchema get address {
    return AddressSchema(getValue<Map<String, dynamic>>('address')!);
  }

  Map<String, Object?> get metadata {
    final map = toMap();
    final knownFields = {'email', 'name', 'age', 'password', 'address'};
    return Map.fromEntries(
        map.entries.where((e) => !knownFields.contains(e.key)));
  }

  @override
  User toModel() {
    if (!isValid) {
      throw AckException(getErrors()!);
    }
    return User(
      email: email,
      name: name,
      age: age,
      password: password,
      address: address.toModel(),
      metadata: metadata,
    );
  }

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for Address
class AddressSchema extends SchemaModel<Address> {
  AddressSchema([Object? value = null]) : super(value);

  static final ObjectSchema schema = Ack.object(
    {
      'street': Ack.string.notEmpty(),
      'city': Ack.string.notEmpty(),
      'zip': Ack.string.nullable(),
    },
    required: ['street', 'city'],
    additionalProperties: false,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<Address, AddressSchema>(
      (data) => AddressSchema(data),
    );
  }

  @override
  AckSchema getSchema() => schema;

  String get street => getValue<String>('street')!;

  String get city => getValue<String>('city')!;

  String? get zip => getValue<String>('zip');

  @override
  Address toModel() {
    if (!isValid) {
      throw AckException(getErrors()!);
    }
    return Address(
      street: street,
      city: city,
      zip: zip,
    );
  }

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}
